Fixed gravity block placement order and snow layer

aria1th opened the pull request.

aria1th committed:
Update README.md

add TODO list

aria1th committed:
Sea pickles go brrrr

Yes it would place sea pickles correctly too.

aria1th committed:
remove sea pickle from todo list

aria1th committed:
Follow Render Layer range brr

aria1th committed:
Update README.md

aria1th committed:
Don't place gravity blocks first

to prevent block falling

aria1th committed:
Replaceable snow block

uhh using context would be better but idk how

aria1th committed:
bump version to 1.0.2

Version differences : Fixed sea pickle count, Render layer reflection, Gravity block placement later, ignore snow block.

aria1th commented:
Gravity block check calls for offset schematic block but not used: its for breaking option because there's possible chance that wrong block is there and break downward block to make gravity blocks fall.

I currently failed to build and somehow this pull request contains same patches so maybe will be fixed soon

jensvh commented:
I'm looking forward to it.

aria1th committed:
Avoid breaking water and lava blocks, bedrock TODO

and you can now place blocks in water / lava.

aria1th committed:
When breakblock is true, check if gravity block can be placed

aria1th commented:
to resolve conflict, should I copy-paste codes from your repository to my repository? seems nothing important is conflicting

aria1th commented:
and maybe I should add checks if torches can be placed on 'water' or 'lava' too... oof

jensvh commented:
> to resolve conflict, should I copy-paste codes from your repository to my repository? seems nothing important is conflicting

They'll be resolved when they get merged.

aria1th committed:
Also don't place torch on top of liquid

but Lava cauldron, Water cauldron can be affected? well torch can't be placed on them anyway.

aria1th commented:
it now works!
Gravity blocks will not be placed when it has blocks to break, or does not have blocks downward. 
Torches will not be placed in wrong places (most likely from ice melting)
Blocks will be placed in water/lava somehow(dangerous, should be controlled by option).
        -its dangerous because original easyplace code uses offset from block origin, so maybe it can be placed at wrong places if there's no facing block. It will be better when it can be replaced with 'CheckReplaceable' code.
Snow Layer block will be considered as air, because you can. 

To be considered:
Added block breaking line, because torches/plants just breaks with one tick and its much faster. but direct interaction can be detected by anticheats, so paper option check should be added.

aria1th commented:
I recently added functionality to use cactus if flippincactus is enabled but only tested in creative. Would it be better to merge and push together?

jensvh commented on README.md:
```diff
@@ -40,3 +40,8 @@ Andrews is the one who made the litematica printer implimentation, I just conver
 This template is available under the CC0 license. Feel free to learn from it and incorporate it in your own projects.
+
+## TODO List (most possible to least possible)
+Breaking based by schematic verifier...maybe.
```
I don't really understand what you mean with this?

aria1th replied:
currently it tries to break everything so it does not bypass bedrock, water, lava thing. also extra blocks too.

jensvh commented on src/main/java/io/github/eatmyvenom/litematicin/utils/Printer.java:
```diff
@@ -373,7 +381,8 @@ public static ActionResult doPrinterAction(MinecraftClient mc) {
                                 }
                         	} else { // For survival
-                            	breaker.startBreakingBlock(pos, mc);
+				mc.interactionManager.attackBlock(pos, Direction.DOWN); //yes, this seemingly needless line adds functionality but paper would not allow it.
```
If we attack the block, there is no need to also break it. Also i think it might be good to check for blocks that can be broken in one tick and add them to the if statement.
```suggestion
				if (!EASY_PLACE_MODE_PAPER.getBooleanValue()) {
				    mc.interactionManager.attackBlock(pos, Direction.DOWN); //yes, this seemingly needless line adds functionality but paper would not allow it.
				    return ActionResult.SUCCESS;
				}
```

jensvh commented on src/main/java/io/github/eatmyvenom/litematicin/utils/Printer.java:
```diff
@@ -495,10 +505,40 @@ public static ActionResult doPrinterAction(MinecraftClient mc) {
                         if (stateSchematic == stateClient) {
                             continue;
-                        }
+                        } else if (sBlock instanceof SandBlock) {
+			BlockPos Offsetpos = new BlockPos(x, y-1, z);
+                    		BlockState OffsetstateSchematic = world.getBlockState(Offsetpos);
+                    		BlockState OffsetstateClient = mc.world.getBlockState(Offsetpos);
+                                        if (OffsetstateClient.isAir() || (breakBlocks && !OffsetstateClient.getBlock().getName().equals(OffsetstateSchematic.getBlock().getName())) )
+                                            continue;
+                                    } else if (sBlock instanceof ConcretePowderBlock) {
+			BlockPos Offsetpos = new BlockPos(x, y-1, z);
+                    		BlockState OffsetstateSchematic = world.getBlockState(Offsetpos);
+                    		BlockState OffsetstateClient = mc.world.getBlockState(Offsetpos);
+                                        if (OffsetstateClient.isAir() || (breakBlocks && !OffsetstateClient.getBlock().getName().equals(OffsetstateSchematic.getBlock().getName())) )
+                                            continue;
+                                    } else if (sBlock instanceof GravelBlock) {
+			BlockPos Offsetpos = new BlockPos(x, y-1, z);
+                    		BlockState OffsetstateSchematic = world.getBlockState(Offsetpos);
+                    		BlockState OffsetstateClient = mc.world.getBlockState(Offsetpos);
+                                        if (OffsetstateClient.isAir() || (breakBlocks && !OffsetstateClient.getBlock().getName().equals(OffsetstateSchematic.getBlock().getName())) )
+                                            continue;
+                                    } else if (sBlock instanceof AnvilBlock) {
+			BlockPos Offsetpos = new BlockPos(x, y-1, z);
+                    		BlockState OffsetstateSchematic = world.getBlockState(Offsetpos);
+                    		BlockState OffsetstateClient = mc.world.getBlockState(Offsetpos);
+                                        if (OffsetstateClient.isAir() || (breakBlocks && !OffsetstateClient.getBlock().getName().equals(OffsetstateSchematic.getBlock().getName())) )
+                                            continue;
+                                    } else if (sBlock instanceof DragonEggBlock) {
+			BlockPos Offsetpos = new BlockPos(x, y-1, z);
+                    		BlockState OffsetstateSchematic = world.getBlockState(Offsetpos);
+                    		BlockState OffsetstateClient = mc.world.getBlockState(Offsetpos);
+                                        if (OffsetstateClient.isAir() || (breakBlocks && !OffsetstateClient.getBlock().getName().equals(OffsetstateSchematic.getBlock().getName())) )
+                                            continue;
+                                    }
```
Here you write the same thing multiple times, this fits in one if statement.
```suggestion
                        } else if (sBlock instanceof SandBlock || sBlock instanceof ConcretePowderBlock 
                        || sBlock instanceof GravelBlock || sBlock instanceof AnvilBlock || sBlock instanceof DragonEggBlock) {
			BlockPos Offsetpos = new BlockPos(x, y-1, z);
                    		BlockState OffsetstateSchematic = world.getBlockState(Offsetpos);
                    		BlockState OffsetstateClient = mc.world.getBlockState(Offsetpos);
                                        if (OffsetstateClient.isAir() || (breakBlocks && !OffsetstateClient.getBlock().getName().equals(OffsetstateSchematic.getBlock().getName())) )
                                            continue;
                                    } 
```

jensvh commented on src/main/java/io/github/eatmyvenom/litematicin/utils/Printer.java:
```diff
@@ -534,6 +574,13 @@ public static ActionResult doPrinterAction(MinecraftClient mc) {
                         Block blockSchematic = stateSchematic.getBlock();
+                        if (blockSchematic instanceof TorchBlock) {
+                        BlockPos Offsetpos = new BlockPos(x, y-1, z);
+                        BlockState OffsetstateSchematic = world.getBlockState(Offsetpos);
```
This can be removed, it seems like you never use it.
```suggestion
```

jensvh commented on src/main/java/io/github/eatmyvenom/litematicin/utils/Printer.java:
```diff
@@ -712,16 +769,25 @@ private static boolean canPlaceFace(FacingData facedata, BlockState stateSchemat
+        if (blockClient instanceof SnowBlock && stateClient.get(SnowBlock.LAYERS) <3) {
+                return false;
+        }
+        if (stateClient.isAir() || stateClient.getBlock().getTranslationKey().contains((String) "water") || stateClient.getBlock().getTranslationKey().contains((String) "lava")) // This is a lot simpler than below. But slightly lacks functionality.
```
If i'm correct, your intension is to not place blocks in water and don't try to break them. This line only does the trick for clicking on blocks (e.g. repeaters).

aria1th replied:
placing blocks to replace water/lava requires 1. bypass your breaking code check, 2. bypass isAir()check. so those lines are needed to ignore snow layer(less than 3), and fluid types.

jensvh reviewed (changes requested):
Thanks for implementing those features. But there are some minor tweaks needed for optimization. I've made some comments on pieces of code, could you change them?
Also is it possible to sync your fork with mine, this should remove the edits i've already aproved from your previous pull request.

aria1th committed:
Update src/main/java/io/github/eatmyvenom/litematicin/utils/Printer.java

add Paper check

Co-authored-by: jensvh <hummiejens@hotmail.com>

aria1th committed:
Update src/main/java/io/github/eatmyvenom/litematicin/utils/Printer.java

Co-authored-by: jensvh <hummiejens@hotmail.com>

aria1th committed:
Update src/main/java/io/github/eatmyvenom/litematicin/utils/Printer.java

Co-authored-by: jensvh <hummiejens@hotmail.com>

jensvh committed:
Merge branch 'master' into snapshot-world-height-patch
